// Copyright (c) 2016-present, Facebook, Inc.
// All rights reserved.
//
// This source code is licensed under the BSD-style license found in the
// LICENSE file in the root directory of this source tree. An additional grant
// of patent rights can be found in the PATENTS file in the same directory.

syntax = "proto3";

package magma.lte;

import "orc8r/protos/common.proto";
import "lte/protos/meteringd.proto";
import "lte/protos/policydb.proto";
import "lte/protos/session_manager.proto";
import "lte/protos/subscriberdb.proto";

option go_package = "magma/lte/cloud/go/protos";

message SetupFlowsRequest {
  // List of rules to activate
  repeated ActivateFlowsRequest rules = 1;
  // epoch to prevent outdated setup calls
  uint64 epoch = 2;
}

message SetupFlowsResult {
  enum Result {
    SUCCESS = 0;
    FAILURE = 1;
    OUTDATED_EPOCH = 2;
  }
  Result result = 1;
}

message ActivateFlowsRequest {
  SubscriberID sid = 1;

  // Subscriber session ipv4 address
  string ip_addr = 2;

  // List of static rules obtained from PCRF
  repeated string rule_ids = 3;

  // List of dynamic rules obtained from PCRF
  repeated PolicyRule dynamic_rules = 4;
}

// DeactivateFlowsRequest can be used to deactivate all flows for a subscriber,
// all flows for some rules, or particular rules for a subscriber, depending on
// which parameters are passed. Rule IDs can apply to static rules or dynamic
// rules
// If no rule ids are given, all flows are deactivated
message DeactivateFlowsRequest {
  SubscriberID sid = 1;
  repeated string rule_ids = 2;
}

message RuleModResult {
  string rule_id = 1;
  enum Result {
    SUCCESS = 0;
    PARTIAL_SUCCESS = 1;
    FAILURE = 2;
  }
  Result result = 2;
}

message ActivateFlowsResult {
  repeated RuleModResult static_rule_results = 1;
  repeated RuleModResult dynamic_rule_results = 2;
}

message DeactivateFlowsResult {
}

message FlowRequest {
  FlowMatch match = 1;
  string app_name = 2;
  uint64  bytes_rx = 3;
}

message FlowResponse {
}

// UEMacFlowRequest is used to link a subscriber ID to a MAC address.
// This is used for Carrier WiFi data session establishment
message UEMacFlowRequest {
  SubscriberID sid = 1;

  // UE MAC address
  string mac_addr = 2;
}

message TableAssignment {
  string app_name = 1;
  uint64 main_table = 2;
  repeated uint64 scratch_tables = 3;
}

message AllTableAssignments {
  repeated TableAssignment table_assignments = 1;
}

message SerializedRyuPacket {
  bytes pkt = 1;
  string imsi = 2;
}

message PacketDropTableId {
  int32 table_id = 1;
}

// --------------------------------------------------------------------------
// Pipelined gateway RPC service
// --------------------------------------------------------------------------
service Pipelined {
  // -------------
  // Metering App:
  // -------------

  // Get subscriber metering flows
  rpc GetSubscriberMeteringFlows (magma.orc8r.Void) returns (FlowTable) {}

  // ----------------
  // Enforcement App:
  // ----------------

  // Setup flows for subscribers (used on restarts)
  rpc SetupFlows (SetupFlowsRequest) returns (SetupFlowsResult) {}

  // Activate flows for a subscriber based on predefined flow templates
  rpc ActivateFlows (ActivateFlowsRequest) returns (ActivateFlowsResult) {}

  // Deactivate flows for a subscriber
  rpc DeactivateFlows (DeactivateFlowsRequest) returns (DeactivateFlowsResult) {}

  // Get policy usage stats
  rpc GetPolicyUsage (magma.orc8r.Void) returns (RuleRecordTable) {}

  // --------
  // DPI App:
  // --------

  // Add new dpi flow
  rpc CreateFlow (FlowRequest) returns (FlowResponse) {};

  // Update flow stats
  rpc UpdateFlowStats (FlowRequest) returns (FlowResponse) {};

  // -----------
  // UE MAC App:
  // -----------

  // Add a flow for a subscriber by matching the provided UE MAC address
  rpc AddUEMacFlow (UEMacFlowRequest) returns (FlowResponse) {}


  // ------------------
  // Packet tracer App:
  // ------------------

  // Initiate a packet send through the switch which will report the
  // table_id that caused a failure in the end
  rpc TracePacket (SerializedRyuPacket) returns (PacketDropTableId) {}


  // --------
  // Debugging:
  // --------

  // Get the flow table assignment for all apps ordered by main table number
  // and name
  rpc GetAllTableAssignments (magma.orc8r.Void) returns (AllTableAssignments);
}
